packages/contracts/contracts/plugin/PluginSetupProcessor.sol

```diff
@@ -186,394 +247,394 @@ contract PluginSetupProcessor is DaoAuthorizable {
 
-        return (permissions, initData);
+        // Avoid stack too deep.
+        emitPrepareUpdateEvent(_dao, newSetupId, _params, preparedDependency, initData);
+
+        return (initData, preparedDependency);
     }
 
     /// @notice Applies the permissions of a prepared update of an UUPS upgradeable contract to a DAO.
     /// @param _dao The address of the updating DAO.
-    /// @param _plugin The address of the `PluginUUPSUpgradeable` proxy contract.
-    /// @param _pluginSetup The address of the `PluginSetup` contract.
-    /// @param _pluginSetupRepo The repository storing the `PluginSetup` contracts of all versions of a plugin.
-    /// @param _initData The initialization data to be passed to the upgradeable plugin contract via `upgradeToAndCall`. // revisit
-    /// @param _permissions The list of multi-targeted permission operations to apply to the updating DAO.
-    function applyUpdate(
-        address _dao,
-        address _plugin,
-        address _pluginSetup,
-        PluginRepo _pluginSetupRepo,
-        bytes memory _initData,
-        PermissionLib.ItemMultiTarget[] calldata _permissions
-    ) external canApply(_dao, APPLY_UPDATE_PERMISSION_ID) {
-        bytes32 setupId = _getSetupId(_dao, _pluginSetup, address(_pluginSetupRepo), _plugin);
-
-        if (updatePermissionHashes[setupId] != _getPermissionsHash(_permissions)) {
-            revert PermissionsHashMismatch();
+    /// @param _params The struct containing the parameters for the `applyInstallation` function.
+    function applyUpdate(address _dao, ApplyUpdate calldata _params)
+        external
+        canApply(_dao, APPLY_UPDATE_PERMISSION_ID)
+    {
+        bytes32 pluginId = _getPluginId(_dao, _params.plugin);
+
+        PluginState storage pluginState = states[pluginId];
+
+        bytes32 setupId = _getSetupId(
+            _params.pluginSetupRef,
+            pHash(_params.permissions),
+            _params.helpersHash,
+            _params.initData,
+            PreparationType.Update
+        );
+
+        // If the plugin block number exceeds the setupId preparation block number,
+        // This means applyUpdate was already called on another setupId
+        // and all the rest setupIds should become idle or setupId is not prepared before.
+        if (pluginState.blockNumber >= pluginState.setupIds[setupId]) {
+            revert SetupNotApplicable(setupId);
         }
 
-        address currentImplementation = PluginUUPSUpgradeable(_plugin).getImplementationAddress();
-        address newImplementation = PluginSetup(_pluginSetup).getImplementationAddress();
+        // Once the applyUpdate is called and arguments are confirmed(including initData)
+        // we update the setupId with the new versionTag, the current helpers. All other
+        // data can be put with bytes(0) as they don't need to be confirmed in the later
+        // prepareUpdate/prepareUninstallation.
+        bytes32 newSetupId = _getSetupId(
```

Comment:
If the two setup ID's are differentiated between the "theoretical" and the "actual" setup ID's, then let's call it `actualSetupId` and not "new", since we're not creating anything.

Reply:
In my opinion, every apply function creates a new setupId which becomes currentSetupId. 

```
pluginState.currentSetupId = newSetupId;
```

From this, it's clear that apply creates a newSetupId which replaces the current one.

Reply:
Oh ok, so this is not to do the integrity check, but rather to keep track of the one in progress. Thank you

Feel free to resolve

Reply:
Ok, my point is about the relationship between the 2. If what we are doing here is not comparing theoretical vs actual, then this comment doesn't apply. But still I don't get why we compute the "old" setupId.

Feel free to resolve if I'm lacking context here